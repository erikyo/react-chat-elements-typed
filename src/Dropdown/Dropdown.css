.rce-dropdown-container {
    position: relative;
}

.rce-dropdown {
    width: 100%;
    box-sizing: border-box;
    padding: 8px 15px;
    background: #fff;
    border-radius: 5px;
    display: none;
    align-items: center;
    justify-content: center;
    transform: scale(0);
    position: absolute;
    box-shadow: 0 0 5px 0 rgb(163 163 163 / 100%);
    transform-origin: left top;
    z-index: 99999;
}

.rce-dropdown-title {
    font-size: 20px;
    font-weight: 500;
}

.rce-dropdown.dropdown-show {
    animation: dropdown-scaling 0.2s ease forwards;
    display: flex;
}

@keyframes dropdown-scaling {
    0% {
        opacity: 0;
    }

    50% {
        opacity: 0.5;
        transform: scale(1.1);
    }

    100% {
        opacity: 1;
        transform: scale(1);
    }
}

.rce-dropdown.dropdown-hide {
    animation: dropdown-reverse-scaling 0.2s ease forwards;
    display: flex;
}

@keyframes dropdown-reverse-scaling {
    0% {
        opacity: 1;
        transform: scale(1);
    }

    50% {
        opacity: 0.5;
        transform: scale(1.1);
    }

    100% {
        opacity: 0;
        transform: scale(0);
    }
}

.rce-dropdown-open__northwest {
    transform-origin: left top;
    left: 0;
    top: 100%;
    margin-top: 5px;
}

.rce-dropdown-open__northeast {
    transform-origin: right top;
    right: 0;
    top: 100%;
    margin-top: 5px;
}

.rce-dropdown-open__southwest {
    transform-origin: left bottom;
    left: 0;
    bottom: 100%;
    margin-bottom: 5px;
}

.rce-dropdown-open__southeast {
    transform-origin: right bottom;
    right: 0;
    bottom: 100%;
    margin-bottom: 5px;
}

.rce-dropdown ul {
    list-style: none;
    margin: 0;
    padding: 0;
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
}

.rce-dropdown ul li {
    white-space: nowrap;
    color: #767676;
    padding: 8px;
    cursor: pointer;
    font-size: 16px;
    width: 100%;
    border-bottom: 1px solid #e9e9e9;
    box-sizing: border-box;
    user-select: none;
}

.rce-dropdown ul li:last-child {
    border: none;
}

.rce-dropdown ul li:hover a {
    color: #3a6d8c;
}

.rce-dropdown.fade {
    opacity: 0;
    transform: scale(1);
    animation: dropdown-fade 0.5s ease forwards;
}

@keyframes dropdown-fade {
    0% {
        opacity: 0;
    }

    100% {
        opacity: 1;
    }
}
