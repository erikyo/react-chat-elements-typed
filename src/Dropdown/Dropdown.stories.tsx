import type { StoryObj } from "@storybook/react";

import Dropdown from "./Dropdown";
import type { IDropdownProps } from "../types";
import { generateDropDownItem } from "../stories/utils";
import { IoOpenOutline } from "react-icons/io5";

const meta = {
	component: Dropdown,
};

export default meta;

type Story = StoryObj<typeof meta>;

const Template: IDropdownProps = {
	buttonProps: {
		icon: {
			component: <IoOpenOutline />,
			float: "left",
			size: 22,
		},
		squared: true,
		outlined: true,
		text: "Click to open the Dropdown",
	},
	animationType: "northwest",
	animationPosition: "default",
	items: [
		generateDropDownItem(1),
		generateDropDownItem(2),
		generateDropDownItem(3),
	],
	style: {
		maxWidth: "200px",
	},
};

export const Default = {
	args: Template,
};
