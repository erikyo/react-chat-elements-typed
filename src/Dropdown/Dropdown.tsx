import type { FC } from "react";
import { useState } from "react";
import "./Dropdown.css";
import classNames from "classnames";
import Button from "../Button/Button";
import type { IDropdownProps } from "../types";

const Dropdown: FC<IDropdownProps> = ({
	animationPosition = "nortwest",
	animationType = "default",
	...props
}) => {
	const [show, setShow] = useState<boolean | undefined>(undefined);

	const onBlur = () => {
		if (show === true) setShow(false);
	};

	return (
		<div
			className={classNames("rce-dropdown-container", props.className)}
			onBlur={onBlur}
		>
			{<Button {...props.buttonProps} onClick={() => setShow(!show)} />}
			<div
				className={classNames(
					"rce-dropdown",
					animationType,
					`rce-dropdown-open__${animationPosition}`,
					{ "dropdown-hide": show === false },
					{ "dropdown-show": show === true },
				)}
			>
				<ul>
					{props.title && (
						<span className="rce-dropdown-title">{props.title}</span>
					)}
					{props.items?.map((x, i) => (
						<li
							key={i.toString()}
							onMouseDown={(e) =>
								props.onSelect instanceof Function
									? props.onSelect(e, i)
									: void 0
							}
						>
							{typeof x !== "string" ? (
								x.icon ? (
									<span className="rce-button-icon--container">
										{(x.icon.float === "right" || !x.icon.float) && (
											<button type={"button"} className={"rce-button-link"}>
												{x.text}
											</button>
										)}

										<span
											style={{
												float: x.icon.float ?? "inherit",
												color: String(x.icon.color) || "inherit",
												fontSize: Number(x.icon.size) || 12,
											}}
											className={classNames(
												"rce-button-icon",
												x.icon.className,
											)}
										>
											{x.icon.component}
										</span>

										{x.icon.float === "left" && (
											<button type={"button"} className={"rce-button-link"}>
												{x.text}
											</button>
										)}
									</span>
								) : (
									<button type={"button"} className={"rce-button-link"}>
										{x.text}
									</button>
								)
							) : (
								<button type={"button"} className={"rce-button-link"}>
									{x}
								</button>
							)}
						</li>
					))}
				</ul>
			</div>
		</div>
	);
};

export default Dropdown;
