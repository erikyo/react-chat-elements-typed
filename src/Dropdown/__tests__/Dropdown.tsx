import { describe, expect, it } from "vitest";
import React from "react";
import Dropdown from "../Dropdown";
import { render } from "@testing-library/react";

describe("Avatar component", () => {
	it("should render without issues", () => {
		const component = render(<Dropdown />);
		expect(component.container).toBeTruthy();
		expect(component.container).toMatchSnapshot();
	});
});
