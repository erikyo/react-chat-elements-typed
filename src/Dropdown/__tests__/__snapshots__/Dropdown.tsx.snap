// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar component > should render without issues 1`] = `
<div>
  <div
    class="rce-dropdown-container"
  >
    <button
      class="rce-button"
      style="border-width: 2px; cursor: pointer; filter: none; flex-direction: row;"
      type="button"
    />
    <div
      class="rce-dropdown default rce-dropdown-open__northwest"
    >
      <ul />
    </div>
  </div>
</div>
`;
